File: leetcode/0525-contiguous-array/solution.py
from typing import List


class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        # map running (ones - zeros) balance -> earliest index it was seen at
        first_seen = {0: -1}
        balance = 0
        best = 0
        for i, x in enumerate(nums):
            balance += 1 if x == 1 else -1
            if balance in first_seen:
                best = max(best, i - first_seen[balance])
            else:
                first_seen[balance] = i
        return best

File: leetcode/0525-contiguous-array/solution.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        unordered_map<int, int> firstSeen;
        firstSeen[0] = -1;
        int balance = 0;
        int best = 0;
        for (int i = 0; i < (int)nums.size(); i++) {
            balance += nums[i] == 1 ? 1 : -1;
            auto it = firstSeen.find(balance);
            if (it != firstSeen.end()) {
                best = max(best, i - it->second);
            } else {
                firstSeen[balance] = i;
            }
        }
        return best;
    }
};
